### PR TITLE
builder: fix processing of invalid substitution syntax

### LINKED ISSUE #37128: Incomplete Dockerfile variable seems to pass the parser
**Description**
Dockerfiles support the replacement of variables with a Bash-like syntax (such as `${variable:-word}`). One would expect that a missing `}` would cause the builder to complain but it doesn't.

**Steps to reproduce the issue:**
1. Create the following seemingly malformed Dockerfile.
```Dockerfile
FROM scratch
ARG var=${aaa:-bbb
```
2. Build it. It succeeds. Okay. Maybe `var` is just pointing at the raw string `${aaa:-bbb` as is...?
```
$ docker build .
Sending build context to Docker daemon  1.693MB
Step 1/2 : FROM scratch
 --->
Step 2/2 : ARG var=${aaa:-bbb
 ---> Running in 6fce8a055a31
Removing intermediate container 6fce8a055a31
 ---> 079813df835c
Successfully built 079813df835c
```

**Describe the results you received:**
If you run `docker inspect 079813df835c`, the `var` variable is actually `bbb`...?
```
$ docker inspect 079813df835c
[
    {
        "Id": "sha256:079813df835ca2e3e93e24925e4b8945a16035886651d3f0ccb432a4296721c7",
        "RepoTags": [],
        "RepoDigests": [],
        "Parent": "",
        "Comment": "",
        "Created": "2018-05-22T22:43:52.70708892Z",
        "Container": "6fce8a055a314364378975a968099e58f2a256519d98a445b4e4fefe1c34e304",
        "ContainerConfig": {
            "Hostname": "6fce8a055a31",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "/bin/sh",
                "-c",
                "#(nop) ",
                "ARG var=bbb"
            ],
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "DockerVersion": "17.12.0-ce",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": null,
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 0,
        "VirtualSize": 0,
        "GraphDriver": {
            "Data": null,
            "Name": "overlay2"
        },
        "RootFS": {
            "Type": "layers"
        },
        "Metadata": {
            "LastTagTime": "0001-01-01T00:00:00Z"
        }
    }
]
```

**Describe the results you expected:**
I was expecting the build a) to fail or b) to produce `"ARG var=${aaa:-bbb"`

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
$ docker version
Client:
 Version:       17.12.0-ce
 API version:   1.35
 Go version:    go1.9.2
 Git commit:    c97c6d6
 Built: Wed Dec 27 20:05:38 2017
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.0-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.2
  Git commit:   c97c6d6
  Built:        Wed Dec 27 20:12:29 2017
  OS/Arch:      linux/amd64
  Experimental: true
```

**Output of `docker info`:**

```
$ docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 17.12.0-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-96-generic
Operating System: Alpine Linux v3.7 (containerized)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.4GiB
Name: node1
ID: CMW4:R5L3:R3JK:XY45:5C6N:FP5O:3S5M:7S4J:RHPM:EG26:CT3S:W454
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 20
 Goroutines: 34
 System Time: 2018-05-22T22:50:03.301170942Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.1
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Tested online with [Play with Docker](https://labs.play-with-docker.com/).

### PR DESCRIPTION
**Pull request by thaJeztah**
fixes https://github.com/moby/moby/issues/37128

The builder did not detect syntax errors in substitutions in the
Dockerfile, causing those values to be processed incorrectly instead of
producing an error.

### Example 1: missing `}`

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${aaa:-bbb
    RUN echo $var
    EOF

#### Before:

    Step 3/3 : RUN echo $var
     ---> Running in f06571e77146
    bbb

#### After:

    Step 2/3 : ARG var=${aaa:-bbb
    failed to process "${aaa:-bbb": syntax error: missing '}'

### Example 2: missing closing `}`, no default value

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${aaa
    RUN echo $var
    EOF

#### Before:

    Step 2/3 : ARG var=${aaa
    failed to process "${aaa": missing ':' in substitution

#### After:

    Step 2/3 : ARG var=${aaa
    failed to process "${aaa": syntax error: missing '}'

### Example 3: double opening bracket (`{`)

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${{aaa:-bbb}
    RUN echo $var
    EOF

#### Before:

    Step 2/3 : ARG var=${{aaa:-bbb}
    failed to process "${{aaa:-bbb}": missing ':' in substitution

#### After:

    Step 2/3 : ARG var=${{aaa:-bbb}
    failed to process "${{aaa:-bbb}": syntax error: bad substitution

### Example 4: double opening bracket (`{`), no default value

    docker build --no-cache -<<'EOF'
    FROM busybox
    ARG var=${{aaa}
    RUN echo $var
    EOF

#### Before:

    Step 2/3 : ARG var=${{aaa}
    failed to process "${{aaa}": missing ':' in substitution

#### After:

    Step 2/3 : ARG var=${{aaa}
    failed to process "${{aaa}": syntax error: bad substitution

**- Description for the changelog**

```Markdown
* Fix parsing of invalid environment variable substitution 
```

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @duglin PTAL

**Review by vdemeester — APPROVED** [review]
LGTM :tiger:

**Comment by thaJeztah** [comment]
@rcjsuen  spotted one more omission in the validation; https://github.com/moby/moby/issues/37128#issuecomment-391734790 - let me add an extra commit to cover that as well

File: builder/dockerfile/shell/envVarTest [hunk 1]
```diff
@@ -119,3 +119,22 @@ A|안녕${XXX:-\$PWD:}xx       |     안녕$PWD:xx
...
+A|${aaa:-bbb ${foo}         |     error
+A|${aaa:-bbb {foo}          |     bbb {foo
+A|${:-bbb}                  |     error
+A|${:+bbb}                  |     error
+A|${-bbb}                   |     error
+A|${+bbb}                   |     error
+A|${-:bbb}                  |     error
+A|${+:bbb}                  |     error
+A|${+}                      |     error
+A|${-}                      |     error
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply builder/dockerfile/shell/envVarTest]: Should already be caught, but I just added (never hurts)

**Comment by thaJeztah** [comment]
Working on some changes, after discussing this on slack.

File: builder/dockerfile/shell/lex.go [hunk 2]
```diff
@@ -261,6 +261,31 @@ func (sw *shellWord) processDollar() (string, error) {
 	}
 
 	sw.scanner.Next()
+
+	if sw.scanner.Peek() == ':' || sw.scanner.Peek() == '-' || sw.scanner.Peek() == '+' {
+		// Invalid ${:xx}, ${-xx}, and ${-} cases
+		//
+		// Note: validity of ${-:bbb} and ${-} is inconsistent between implementations.
+		// Treat it as invalid, because we cannot substitute the `-` with
+		// anything useful.
```
Comment (ijc) [inline builder/dockerfile/shell/lex.go]: Why is `$-` special but `$*` `$@` `$#` `$?` `$$` `$0` and `$_` (other such special parameters in POSIX) do not appear to be?

I also think it is a mistake to conflate the existence/behaviour of those special parameters with the entirely unrelated support for the `${parameter:offset}` syntax, which is non-POSIX and hence not supported by dash but is supported by `bash` and whatever the alpine/busybox shell is these days (`ash`?).

Don't be confused into thinking that `${-:offset}` is somehow making `-:` into something special, it is not, it's just two unrelated concepts being used at the same time.

I think you should decide whether or not the `${parameter:offset}` form is valid, irrespective of whether `parameter` is `-` or `foo` or `PATH` and just implement what you chose there.

Then, completely separately, decide which, if any, special parameters are supported. This could be as simple as having no way to set `-` or `@` etc and therefore treating them as undefined variables at this level or could be done by seeing those names, whenever they are used, and rejecting them.

BTW I don't think `${:xx}` is valid anywhere, is it? Because it is missing the parameter name entirely so it is just syntactically invalid.

**Comment by thaJeztah** [comment]
@ijc updated PTAL; let me know what you think 👍 👎

**Review by tianon — APPROVED** [review]
That is a lot of new edge case tests! :heart_eyes:

**Comment by thaJeztah** [comment]
They're unit-tests so didn't break the bank :joy:

**Review by ijc — APPROVED** [review]

**Comment by thaJeztah** [comment]
ping @tonistiigi in case you need changes in BuildKit for this 👍

**Comment by thaJeztah** [comment]
> buildkit uses the same package

d'oh! I looked for that last week, forgot I did 😂

**Review by vieux — APPROVED** [review]
LGTM @tonistiigi I'll let you merge since you own the builder.
